### PR TITLE
chore & fix(vue-virtual): fix package information & possible memory leak 

### PR DESCRIPTION
**Pull request by salqueng**
chore(vue-virtual)
- upgrade virtual-core dependency to 3.0.0-beta.23
- update tsconfig and script config files to correctly map vue-virtual.

fix(vue-virtual)
- fix possible memory leak by calling doClean during unmounting.

### CODE REVIEW

File: packages/vue-virtual/src/index.ts [hunk 1]
```diff
@@ -54,6 +62,10 @@ function useVirtualizerBase<TScrollElement, TItemElement extends Element>(
     },
   )
 
+  onUnmounted(() => {
+    doClean?.()
+  })
```
Comment (s-montigny-desautels) [inline packages/vue-virtual/src/index.ts:67]: We should probably use `onScopeDispose` for this.

File: packages/vue-virtual/src/index.ts [hunk 2]
```diff
@@ -19,25 +27,25 @@ function useVirtualizerBase<TScrollElement, TItemElement extends Element>(
...
-        onChange: (instance) => {
-          // Force an update event
-          triggerRef(state)
-          options.onChange?.(instance)
-        },
-      })
+  const applyOptions = (
+    options: VirtualizerOptions<TScrollElement, TItemElement>,
+  ) => {
+    doClean?.()
```
Comment (s-montigny-desautels) [inline packages/vue-virtual/src/index.ts:35]: I analysed other adapter and the core lib and I don't think we need to call doClean on each applyOptions call. It should only be called on unMounted for cleanup. What do you think?

File: packages/vue-virtual/src/index.ts [hunk 3]
```diff
@@ -19,25 +27,25 @@ function useVirtualizerBase<TScrollElement, TItemElement extends Element>(
...
+      },
+    })
 
-      virtualizer._willUpdate()
-      virtualizer.measure()
-      doClean = virtualizer._didMount()
-    }
-  })()
+    virtualizer._willUpdate()
+    virtualizer.measure()
```
Comment (s-montigny-desautels) [inline packages/vue-virtual/src/index.ts:46]: Is this still necessary? There is a `ResizeObserver` watching for the scrollElement size change. I'm not having any issues when the initial size of the scrollElement is at 0 on the first render without it.

**Review by s-montigny-desautels — COMMENTED** [review]
Hey thanks for this! 
I have added some comments on your changes. Also, I don't think we should include the `package-lock.json` and the `yarn.lock` files in the PR. Can you remove them?

**Comment by s-montigny-desautels** [comment]
I've experienced a bit more with the structure of the code for the Vue adapter. I think it can be simpler. Here are the `useVirtualizerBase` function that I propose.

```ts
function useVirtualizerBase<TScrollElement, TItemElement extends Element>(
  options: MaybeRef<VirtualizerOptions<TScrollElement, TItemElement>>,
): Ref<Virtualizer<TScrollElement, TItemElement>> {
  const virtualizer = new Virtualizer(unref(options))
  const state = shallowRef(virtualizer)

  const cleanup = virtualizer._didMount()

  watch(
    () => unref(options).getScrollElement(),
    (el) => {
      if (el) {
        virtualizer._willUpdate()
      }
    },
    {
      immediate: true,
    },
  )

  watch(
    () => unref(options),
    (options) => {
      virtualizer._willUpdate()
      virtualizer.setOptions({
        ...options,
        onChange: (instance) => {
          triggerRef(state)
          options.onChange?.(instance)
        },
      })
    },
    {
      immediate: true,
    },
  )

  onScopeDispose(cleanup)

  return state
}
```

The first watch only calld the `_willUpdate` function. It's the only function needed to be called when the `scrollElement` change. The second watch update the options in the same way we are already doing it.

I've also applied the changes that I have commented. What do you think?

**Comment by salqueng** [comment]
> I've experienced a bit more with the structure of the code for the Vue adapter. I think it can be simpler. Here are the `useVirtualizerBase` function that I propose.
> 
> ```ts
> function useVirtualizerBase<TScrollElement, TItemElement extends Element>(
>   options: MaybeRef<VirtualizerOptions<TScrollElement, TItemElement>>,
> ): Ref<Virtualizer<TScrollElement, TItemElement>> {
>   const virtualizer = new Virtualizer(unref(options))
>   const state = shallowRef(virtualizer)
> 
>   const cleanup = virtualizer._didMount()
> 
>   watch(
>     () => unref(options).getScrollElement(),
>     (el) => {
>       if (el) {
>         virtualizer._willUpdate()
>       }
>     },
>     {
>       immediate: true,
>     },
>   )
> 
>   watch(
>     () => unref(options),
>     (options) => {
>       virtualizer._willUpdate()
>       virtualizer.setOptions({
>         ...options,
>         onChange: (instance) => {
>           triggerRef(state)
>           options.onChange?.(instance)
>         },
>       })
>     },
>     {
>       immediate: true,
>     },
>   )
> 
>   onScopeDispose(cleanup)
> 
>   return state
> }
> ```
> 
> The first watch only calld the `_willUpdate` function. It's the only function needed to be called when the `scrollElement` change. The second watch update the options in the same way we are already doing it.
> 
> I've also applied the changes that I have commented. What do you think?

https://stackblitz.com/edit/vitejs-vite-ytsidg
I made a demo with your suggestion. I also added "load row" button to add items dynamically.

When I press the load row button, only 5 items are visible and other items are not visible. I'm not sure what is the problem.

**Comment by s-montigny-desautels** [comment]
> https://stackblitz.com/edit/vitejs-vite-ytsidg I made a demo with your suggestion. I also added "load row" button to add items dynamically.
> 
> When I press the load row button, only 5 items are visible and other items are not visible. I'm not sure what is the problem.

It seems to only be an issue with Firefox. Moving the `_willUpdate` call after the `setOptions` call fixes this issue.
I did dig a little bit more and I found out that the `ResizeObserver` behaviour is different on Firefox and Chrome. On Chrome, the ResizeObserver callback is called when we add the child node to the element, but not on Firefox.

Anyhow, I've updated your [example](https://stackblitz.com/edit/vitejs-vite-ldoeok?file=src%2Fcomposable%2Fvirtual.ts,src%2FApp.vue).

**Comment by salqueng** [comment]
Your code definitely works! It is much simple and more performant than previous commits 👍. I've almost forgot about immediate watch and onScopeDispose!

However, this may not be the release blocker, but I couldn't get my private use cases with your new vue adapter - prepending items and keeping scroll position during scrolling. It worked when I tested with the previous adapter. I think calling measure method made a difference.

I want to create and show you the stackblitz example, but it requires some jobs so I'll post it in this weekend. This issue might related to https://github.com/TanStack/virtual/discussions/431

**Comment by s-montigny-desautels** [comment]
Hey! I've been able to reproduce your use case. I've also tried using the React adapter and we have the same behaviour. The scroll position is not kept when scrolling.

I think we should wait for a fix in @tantack/virtual-core for this. 

I will merge your changes and will apply my proposed change to the adapter. Thanks for your help!

**Pull request merged**